include/ignition/math/detail/Capsule.hh

```diff
@@ -0,0 +1,157 @@
+ * limitations under the License.
+ *
+*/
+#ifndef IGNITION_MATH_DETAIL_CAPSULE_HH_
+#define IGNITION_MATH_DETAIL_CAPSULE_HH_
+
+#include <ignition/math/Inertial.hh>
+
+namespace ignition
+{
+namespace math
+{
+
+//////////////////////////////////////////////////
+template<typename T>
+Capsule<T>::Capsule(const T _length, const T _radius)
+{
+  this->length = _length;
+  this->radius = _radius;
+}
+
+//////////////////////////////////////////////////
+template<typename T>
+Capsule<T>::Capsule(const T _length, const T _radius, const Material &_mat)
+{
+  this->length = _length;
+  this->radius = _radius;
+  this->material = _mat;
+}
+
+//////////////////////////////////////////////////
+template<typename T>
+T Capsule<T>::Radius() const
+{
+  return this->radius;
+}
+
+//////////////////////////////////////////////////
+template<typename T>
+void Capsule<T>::SetRadius(const T _radius)
+{
+  this->radius = _radius;
+}
+
+//////////////////////////////////////////////////
+template<typename T>
+T Capsule<T>::Length() const
+{
+  return this->length;
+}
+
+//////////////////////////////////////////////////
+template<typename T>
+void Capsule<T>::SetLength(const T _length)
+{
+  this->length = _length;
+}
+
+//////////////////////////////////////////////////
+template<typename T>
+const Material &Capsule<T>::Mat() const
+{
+  return this->material;
+}
+
+//////////////////////////////////////////////////
+template<typename T>
+void Capsule<T>::SetMat(const Material &_mat)
+{
+  this->material = _mat;
+}
+
+//////////////////////////////////////////////////
+template<typename T>
+bool Capsule<T>::operator==(const Capsule &_capsule) const
+{
+  return equal(this->radius, _capsule.Radius()) &&
+    equal(this->length, _capsule.Length()) &&
+    this->material == _capsule.Mat();
+}
+
+//////////////////////////////////////////////////
+template<typename T>
+bool Capsule<T>::MassMatrix(MassMatrix3<T> &_massMat) const
+{
+  // mass and moment of inertia of cylinder about centroid
+  MassMatrix3<T> cylinder;
+  cylinder.SetFromCylinderZ(this->material, this->length, this->radius);
+
+  // mass and moment of inertia of hemisphere about centroid
+  const T r2 = this->radius * this->radius;
+  const T hemisphereMass = this->material.Density() *
+      2. / 3. * IGN_PI * r2 * this->radius;
+  // efunda.com/math/solids/solids_display.cfm?SolidName=Hemisphere
+  const T ixx = 83. / 320. * hemisphereMass * r2;
+  const T izz = 2. / 5.  * hemisphereMass * r2;
+  MassMatrix3<T> hemisphere(hemisphereMass, Vector3<T>(ixx, ixx, izz),
+                            Vector3<T>::Zero);;
+
+  // Distance from centroid of cylinder to centroid of hemisphere,
+  // since centroid of hemisphere is 3/8 radius from its flat base
+  const T dz = this->length / 2. + this->radius * 3. / 8.;
+  Inertial<T> upperCap(hemisphere, Pose3<T>(0, 0, dz, 0, 0, 0));
+  Inertial<T> lowerCap(hemisphere, Pose3<T>(0, 0, -dz, 0, 0, 0));
```

Comment:
I'm missing a 180 degree rotation here, but I guess it wouldn't make a difference since this is about the centroid?

Reply:
yeah, it doesn't make a difference because it's about the centroid